pyserver/main.py

```diff
@@ -0,0 +1,362 @@
+    topic_total = 0
+    topic_list = {}
+    for subtopic, subtopic_data in topic_data["subtopics"].items():
+      topic_total += subtopic_data["total"]
+      if subtopic in nested_claims:
+        # this one has some dupes
+        # for each duplicate claim, we list the duplicate ids
+        # but we don't currently merge them as "similar claims"
+        rerank = {}
+        for c in subtopic_data["claims"]:
+          if c in dupe_counts:
+            new_label = c + " (" + str(len(dupe_counts[c]) + 1) + "x:"
+            for ckey in dupe_counts[c]:
+              new_label += " " + str(ckey) + ","
+            new_label += ")"
+            rerank[new_label] = len(dupe_counts[c])
+          else:
+            rerank[c] = 0
+
+        # sort the most-duplicated to the top
+        ranked = sorted(rerank.items(), key=lambda x: x[1], reverse=True)
+        new_claims = [r[0] for r in ranked]
+        topic_list[subtopic] = {"total" : subtopic_data["total"], "claims" : new_claims}
+      else:
+        topic_list[subtopic] = {"total" : subtopic_data["total"], "claims" : subtopic_data["claims"]}
+
+    # sort subtopics themselves
+    sorted_topics = sorted(topic_list.items(), key=lambda x: x[1]["total"], reverse=True)
+
+    sorted_tree[topic] = {"total" : topic_total, "topics" : sorted_topics}
+
+  # sort full tree
+  full_sort_tree = sorted(sorted_tree.items(), key=lambda x: x[1]["total"], reverse=True)
+ 
+  if log_to_wandb:
+    wandb.init(project = config.WANDB_PROJECT_NAME,
+               config = {"model" : config.MODEL})
+    report_data = [[json.dumps(full_sort_tree, indent=2)]]
+    wandb.log({
+      "u/4/N_tok" : TK_TOT,
+      "u/4/in_tok": TK_IN,
+      "u/4/out_tok" : TK_OUT,
+      "dedup_subclaims" : wandb.Table(data=dupe_logs, columns = ["sub_claim_list", "deduped_claims"]),
+      "t3c_report" : wandb.Table(data=report_data, columns = ["t3c_report"])
+      # TODO: do we need this?
+      #"num_claims" : NUM_CLAIMS,
+      #"num_topics_post_sort" : NUM_TOPICS_STEP_3
+    })
+  return {"tree" : full_sort_tree, "per_topic_dupes" : nested_claims, "dupe_counts" : dupe_counts}
+
+#TODO: refactor into separate testing script
+
+######################
+# Testing the server #
+#--------------------#
+
+sample_tree_4o = {'tree': {'taxonomy': [{'topicName': 'Pets', 'topicShortDescription': 'General opinions about common household pets.', 'subtopics': [{'subtopicName': 'Cats', 'subtopicShortDescription': 'Positive sentiments towards cats.'}, {'subtopicName': 'Dogs', 'subtopicShortDescription': 'Positive sentiments towards dogs.'}, {'subtopicName': 'Birds', 'subtopicShortDescription': 'Uncertainty or mixed feelings about birds.'}]}]}, 'usage': {'completion_tokens': 125, 'prompt_tokens': 220, 'total_tokens': 345, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}}
```

Comment:
accidental duplicate of test_pipeline.py? or intentionally also here?

Reply:
intentional for now—pytest only CHECKS, and doesn't easily SHOW output; need to set up something in parallel that does just show the output for a bunch of calls